src/workerd/api/node/url.c++

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+#include "url.h"
+
+#include "ada.h"
+
+namespace workerd::api::node {
+kj::ArrayPtr<const char> UrlUtil::domainToASCII(kj::String domain) {
+  if (domain.size() == 0) {
+    return {};
+  }
+
+  // It is important to have an initial value that contains a special scheme.
+  // Since it will change the implementation of `set_hostname` according to URL
+  // spec.
+  auto out = ada::parse<ada::url_aggregator>("ws://x");
+  JSG_REQUIRE(out.has_value(), Error, "URL parsing failed"_kj);
+  if (!out->set_hostname({domain.begin(), domain.size()})) {
+    return {};
+  }
+  auto result = out->get_hostname();
+  return kj::ArrayPtr<const char>(result.data(), result.size());
+}
```

Comment:
nit: there is a fair amount of duplication in these two methods. Completely optional but it might make sense to move the bulk of the impl is a separate utility method that is used by both of these, just to eliminate the duplication.